### PR TITLE
[FIX] account: match exact amount when text/partner check fails

### PR DESCRIPTION
**Pull request by agr-odoo**
Create a reconciliation model as follows:
- Type: Rule to match invoices/bills
- Match Invoice/bill with Label: False
- Match Invoice/bill with Note: False
- Match Invoice/bill with Reference: False
- Partner is set: False Create an invoice for $900 and confirm it.

Create a bank statement line for $900, with no partner set.
Go to reconcile the bank statement line.

Issue: The invoice is not automatically suggested

This occurs because there is no check for the exact amount as in previous version

opw-3648260

### CODE REVIEW

**Review by aliyatastemirova — APPROVED** [review]
tested functionally and it works.
we had it like that in 15.0, so I assume it is still the desired behaviour.
if I were to nitpick, I'd add newlines before and after the `if not partner:` block for readability but that's just me.

File: addons/account/models/account_reconcile_model.py [hunk 1]
```diff
@@ -812,19 +812,22 @@ def _get_invoice_matching_amls_candidates(self, st_line, partner):
...
+        else:
+            order = 'date_maturity ASC, date ASC, id ASC'
 
-            if self.matching_order == 'new_first':
-                order = 'date_maturity DESC, date DESC, id DESC'
-            else:
-                order = 'date_maturity ASC, date ASC, id ASC'
+        if not partner:
+            aml_domain += [
+                ('balance', '=', -st_line.amount_residual),
```
Comment (smetl) [inline addons/account/models/account_reconcile_model.py:822]: Wrong currency! You are comparing appels with oranges.
- add a check on the same currency
- match on balance or amount_currency depending the currency of the company/st line

Note: the currency of the statement line is given by: `st_line.foreign_currency_id or st_line.journal_id.currency_id or st_line.company_currency_id`

**Review by smetl — CHANGES_REQUESTED** [review]
add a test in foreign currency?

File: addons/account/models/account_reconcile_model.py (hunk 1 above)
Comment (agr-odoo) [reply addons/account/models/account_reconcile_model.py]: Thanks for the heads up!

File: addons/account/models/account_reconcile_model.py [hunk 2]
```diff
@@ -812,19 +812,26 @@ def _get_invoice_matching_amls_candidates(self, st_line, partner):
...
-            amls = self.env['account.move.line'].search(aml_domain, order=order)
-            if amls:
-                return {
-                    'allow_auto_reconcile': False,
-                    'amls': amls,
-                }
+                aml_amount_field = 'amount_currency'
+            aml_domain += [
+                (aml_amount_field, '=', -st_line.amount_residual),
+            ]
```
Comment (smetl) [inline addons/account/models/account_reconcile_model.py:828]:
```suggestion
if not partner:
    st_line_currency = st_line.foreign_currency_id or st_line.journal_id.currency_id or st_line.company_currency_id
    if st_line_currency == self.company_id.currency_id:
        aml_domain = [
            ('currency_id', '=', st_line_currency.id),
            ('amount_residual', '=', -st_line.amount_residual),
        ]
    else:
        aml_domain = [
            ('currency_id', '=', st_line_currency.id),
            ('amount_residual_currency', '=', -st_line.amount_residual),
        ]
```

**Review by smetl — COMMENTED** [review]
It's still nawak

File: addons/account/models/account_reconcile_model.py (hunk 2 above)
Comment (agr-odoo) [reply addons/account/models/account_reconcile_model.py]: Ok for the amount, but the currency is already set by the `match_currency` settings
https://github.com/odoo/odoo/blob/41ca2fcf0449e9b30075808734984b032cb70901/addons/account/models/account_reconcile_model.py#L658-L659

**Comment by agr-odoo** [comment]
@smetl Should be ok now

File: addons/account/models/account_reconcile_model.py (hunk 2 above)
Comment (smetl) [reply addons/account/models/account_reconcile_model.py]: @agr-odoo Only when `match_same_currency` is ticked

Comment (agr-odoo) [reply addons/account/models/account_reconcile_model.py]: If it's not ticked, why do we want to match the currency here?

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**